drivers/spi/spi_pico_pio.c

```diff
@@ -0,0 +1,403 @@
+/*
+ * Copyright (c) 2023 Stephen Boylan <stephen.boylan@beechwoods.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT raspberrypi_pico_spi_pio
+
+#define LOG_LEVEL CONFIG_SPI_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(spi_pico_pio);
+
+#include <zephyr/sys/util.h>
+#include <zephyr/sys/sys_io.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/drivers/pinctrl.h>
+#include "spi_context.h"
+
+#include <zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h>
+
+#include <hardware/pio.h>
+#include "hardware/clocks.h"
+
+#define PIO_CYCLES     (4)
+#define PIO_FIFO_DEPTH (4)
+
+struct spi_pico_pio_config {
+	const struct device *piodev;
+	const struct pinctrl_dev_config *pin_cfg;
+	struct gpio_dt_spec clk_gpio;
+	struct gpio_dt_spec mosi_gpio;
+	struct gpio_dt_spec miso_gpio;
+	const uint32_t clock_freq;
+};
+
+struct spi_pico_pio_data {
+	struct spi_context spi_ctx;
+	uint32_t tx_count;
+	uint32_t rx_count;
+	PIO pio;
+	size_t pio_sm;
+	int dfs;
+};
+
+RPI_PICO_PIO_DEFINE_PROGRAM(spi_cpol_0_cpha_0, 0, 1,
+			    /*     .wrap_target */
+			    0x6101, /*  0: out    pins, 1         side 0 [1] */
+			    0x5101, /*  1: in     pins, 1         side 1 [1] */
+				/*     .wrap */
+);
+
+RPI_PICO_PIO_DEFINE_PROGRAM(spi_cpol_1_cpha_1, 0, 2,
+			    /*     .wrap_target */
+			    0x7021, /*  0: out    x, 1            side 1 */
+			    0xa101, /*  1: mov    pins, x         side 0 [1] */
+			    0x5001, /*  2: in     pins, 1         side 1 */
+				/*     .wrap */
+);
+
+static float spi_pico_pio_clock_divisor(const struct spi_pico_pio_config *dev_cfg,
+	uint32_t spi_frequency)
+{
+	return (float)dev_cfg->clock_freq / (float)(PIO_CYCLES * spi_frequency);
+}
+
+static uint32_t spi_pico_pio_maximum_clock_frequency(const struct spi_pico_pio_config *dev_cfg)
+{
+	return dev_cfg->clock_freq / PIO_CYCLES;
+}
+
+static uint32_t spi_pico_pio_minimum_clock_frequency(const struct spi_pico_pio_config *dev_cfg)
+{
+	return dev_cfg->clock_freq / (PIO_CYCLES * UINT16_MAX);
```

Comment:
Sorry. It my misunderstanding.
The divider allows to set between 1 to 65536, UINT16_MAX is 65536.
Your original code is right.

Reply:
Ah, yes - I should have checked that header file.  Yes, 65536 is the correct value;  I'll push that change.